pymedphys/_experimental/pseudonymisation/strategy.py

```diff
@@ -32,27 +32,32 @@
 
 def get_pseudonymous_replacement_value(keyword, value):
     """
-    Get an appropriate value for a DICOM element based on its VR and
-    take in to account if it is an SOP Class UID.
-    Cache values of keywords that have already come in and re-use
-    the pseudonym for the *value* if it has already been used/replaced.
+    Get an appropriate value for a DICOM element based on its VR
+    that ensures consistent one way mapping to a replacement value
+    which will not reveal Private Health Information
 
     Parameters
     ----------
-    keyword : TYPE
-        DESCRIPTION.
+    keyword : A text string representing which DICOM attribute is to have it's value replaced
+
+    value:  The current value of the DICOM element
+
 
     Returns
     -------
-    None.
+    An appropriate value for a DICOM element based on its VR.
+    Strings are hashed.  Dates are shifted. Ages are jittered.
+    A KeyError will be raised if the VR of the attribute is not addressed by the dispatch table
 
     """
     vr = get_baseline_keyword_vr_dict()[keyword]
-    replacement_value = value
     if vr in pseudonymisation_dispatch:
         replacement_value = pseudonymisation_dispatch[vr](value)
     else:
-        logging.warning("VR of %s not found in pseudonymisation dispatch table", vr)
+        logging.error(
+            "VR of %s for %s not found in pseudonymisation dispatch table", vr, keyword
+        )
+        raise KeyError
```

Comment:
So that the traceback is a more direct link can we do EAFP instead of LBYL:

```python
try:
	replacement_value = pseudonymisation_dispatch[vr](value)
except KeyError:
    logging.error(
        "VR of %s for %s not found in pseudonymisation dispatch table", vr, keyword
    )
    raise
```

Reply:
Or better:

```python
try:
	dispatch = pseudonymisation_dispatch[vr]
except KeyError:
    logging.error(
        "VR of %s for %s not found in pseudonymisation dispatch table", vr, keyword
    )
    raise

replacement_value = dispatch(value)
```

That way, a KeyError in the function call won't be incorrectly caught.

Reply:
But even better still, if you really do want it to be a warning, you can wrap `get_pseudonymous_replacement_value(keyword, value)` in a `try/except`, and then do a logging.warning there. As such, the logging.error potentially doesn't belong in this function, if you intend to turn it into a warning.

Reply:
ok.  no try/catch.  if there's a problem, it's entirely up to the calling code to deal with it, and if not, the stack trace/dump should have the values not only of the vr that the dict didn't support, but also which keyword was the one responsible for it.  I'm used to the paradigm of leaving breadcrumbs, but if EAFP is the preferred approach in python, I'm not going to argue.
This motivates creation of a validation/helper function so that a programmer can test that their combination of keywords and strategy are compatible up front.  Especially if the keywords become user (rather than programmer) defined.

Reply:
> it's entirely up to the calling code to deal with it

Yup, beautiful :) I believe that's the heart of the following:

> ... the emphasis is on "lazy" code: be strict in what you will accept before you begin, and promise as little as possible in return. Remember, if your contract indicates that you'll accept anything and promise the world in return, then you've got a lot of code to write!

> This motivates creation of a validation/helper function so that a programmer can test that their combination of keywords and strategy are compatible up front. Especially if the keywords become user (rather than programmer) defined.

I really like that concept.